src/pcp/ipcs/pcp-ipcs.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2016 Fujitsu.
.\" Copyright (c) 2019 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PCP-IPCS 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-ipcs\f1 \- provide information on IPC facilities
.SH SYNOPSIS
\f3pcp\f1 [\f2pcp\ options\f1] \f3ipcs\f1
[\f3\-luV?\f1] [\f2output\-format\f1]
.SH DESCRIPTION
.B pcp-ipcs
provides information on the inter-process communication facilities
for which the calling process has read access.
.SH OPTIONS
When invoked via the
.BR pcp (1)
command, the
.BR \-h /\c
.BR \-\-host ,
.BR \-a /\c
.BR \-\-archive ,
.BR \-O /\c
.BR \-\-origin ,
.BR \-Z /\c
.BR \-\-timezone
and several other
.I pcp options
become indirectly available; refer to
.BR PCPIntro (1)
for a complete description of these options.
.PP
The additional command line options available for
.B pcp-ipcs
are:
.\" .TP
.\" \fB\-i\fR \fIid\fR, \fB\-\-id\fR=\fIid\fR
.\" Print details only on the resource identified by
.\" .IR id .
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version information and exit.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display the usage message and exit.
.\" .SS "Resource options"
.\" .TP
.\" \fB\-q\fR, \fB\-\-queues\fR
.\" Write information about active message queues.
.\" .TP
.\" \fB\-m\fR, \fB\-\-shmems\fR
.\" Write information about active shared memory segments.
.\" .TP
.\" \fB\-s\fR, \fB\-\-semaphores\fR
.\" Write information about active semaphore sets.
.\" .TP
.\" \fB\-a\fR, \fB\-\-all\fR
.\" Write information about all three resources (default).
.SS "Output formats"
.\" Of these options only one takes effect: the last one specified.
.\" .TP
.\" \fB\-c\fR, \fB\-\-creator\fR
.\" Show creator and owner.
.TP
\fB\-l\fR, \fB\-\-limits\fR
Show resource limits.
.\" .TP
.\" \fB\-p\fR, \fB\-\-pid\fR
.\" Show PIDs of creator and last operator.
.\" .TP
.\" \fB\-t\fR, \fB\-\-time\fR
.\" Write time information.
.\" The time of the last control operation that changed
.\" the access permissions for all facilities, the time of the last
.\" .I msgsnd()
.\" and
.\" .I msgrcv()
.\" operations on message queues, the time of the last
.\" .I shmat()
.\" and
.\" .I shmdt()
.\" operations on shared memory, and the time of the last
.\" .I semop()
.\" operation on semaphores.
.TP
\fB\-u\fR, \fB\-\-summary\fR
Show status summary.
.\" .SS "Representation"
.\" These affect only the \fB\-l\fR (\fB\-\-limits\fR) option.
.\" .TP
.\" \fB\-b\fR, \fB\-\-bytes\fR
.\" Print sizes in bytes.
.\" .TP
.\" .B \-\-human
.\"Print sizes in human-readable format.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.PP
For environment variables affecting PCP tools, see \fBpmGetOptions\fP(3).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pcp (1),
.BR ipcmk (1),
.BR ipcrm (1),
.BR msgrcv (2),
.BR msgsnd (2),
.BR semget (2),
.BR semop (2),
.BR shmat (2),
.BR shmdt (2)
and
.BR shmget (2).

.\" control lines for scripts/man-spell
.\" +ok+ {all of the following are SystemV SHM/SEM libc routines}
.\" +ok+ semget shmget msgrcv msgsnd semop shmat shmdt ipcmk ipcrm
.\" +ok+ ipcs {from pcp-ipcs}


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pcp/ipcs/pcp-ipcs.py
#!/usr/bin/env pmpython
#
# Copyright (C) 2016 fujitsu (wulm.fnst@cn.fujitsu.com).
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=bad-continuation,bad-whitespace
""" Provide information on IPC facilities """

import sys
from pcp import pmapi
from cpmapi import PM_TYPE_U32

class Ipcs(object):
    """ provides information on the inter-process communication
        facilities for which the calling process has read access.
    """

    def __init__(self):
        """ Construct object - prepare for command line handling """
        self.show_limit = 0
        self.show_summary = 0
        self.opts = self.options()
        self.context = None

    def options(self):
        """ Setup default command line argument option handling """
        opts = pmapi.pmOptions()
        opts.pmSetOptionCallback(self.option)
        opts.pmSetShortOptions("luV?")
        opts.pmSetLongOptionHeader("Options")
        # other options will to do(wulm.fnst@cn.fujitsu.com)
        opts.pmSetLongOption("limits", 0, 'l', '', "show resource limits")
        opts.pmSetLongOption("summary", 0, 'u', '', "show status summary")
        opts.pmSetLongOptionVersion()
        opts.pmSetLongOptionHelp()
        return opts

    def option(self, opt, optarg, index):
        """ Perform setup for an individual command line option """
        if opt == 'l':
            self.show_limit = 1
        elif opt == 'u':
            self.show_summary = 1

    def extract(self, descs, result):
        """ Extract the set of metric values from a given pmResult """
        values = []
        for index, desc in enumerate(descs):
            if result.contents.get_numval(index) > 0:
                atom = self.context.pmExtractValue(
                                result.contents.get_valfmt(index),
                                result.contents.get_vlist(index, 0),
                                desc.contents.type, PM_TYPE_U32)
                values.append(int(atom.ul))
            else:
                values.append(int(0))
        return values

    def execute(self):
        """ Using a PMAPI context (could be either host or archive),
            fetch and report a fixed set of values related to ipc.
        """
        metrics_limit = ('ipc.msg.max_msgqid',  'ipc.msg.max_msgsz',
                         'ipc.msg.max_defmsgq', 'ipc.shm.max_seg',
                         'ipc.shm.max_segsz',   'ipc.shm.max_shmsys',
                         'ipc.shm.min_segsz',   'ipc.sem.max_semid',
                         'ipc.sem.max_perid',   'ipc.sem.num_undo',
                         'ipc.sem.max_ops',     'ipc.sem.max_semval')

        metrics_summary = ('ipc.msg.used_queues', 'ipc.msg.tot_msg',
                           'ipc.msg.tot_bytes', 'ipc.shm.used_ids',
                           'ipc.shm.tot',       'ipc.shm.rss',
                           'ipc.shm.swp',       'ipc.shm.swap_attempts',
                           'ipc.shm.swap_successes',
                           'ipc.sem.used_sem',  'ipc.sem.tot_sem')
        pmids_limit = self.context.pmLookupName(metrics_limit)
        descs_limit = self.context.pmLookupDescs(pmids_limit)

        pmids_summary = self.context.pmLookupName(metrics_summary)
        descs_summary = self.context.pmLookupDescs(pmids_summary)

        if self.show_limit == 1:
            result_limit = self.context.pmFetch(pmids_limit)
            values_limit = self.extract(descs_limit, result_limit)
            self.context.pmFreeResult(result_limit)
            self.report(values_limit)

        elif self.show_summary == 1:
            result_summary = self.context.pmFetch(pmids_summary)
            values_summary = self.extract(descs_summary, result_summary)
            self.context.pmFreeResult(result_summary)
            self.report(values_summary)

    def get_pagesize(self):
        metric_pagesize = "hinv.pagesize"
        pmid_pagesize = self.context.pmLookupName(metric_pagesize)
        desc_pagesize = self.context.pmLookupDescs(pmid_pagesize)
        result = self.context.pmFetch(pmid_pagesize)
        pagesize = self.extract(desc_pagesize, result)
        return pagesize[0]

    def report(self, values):
        kb = 1024
        pagesize = self.get_pagesize()
        pagesize_kb = pagesize / kb
        if self.show_limit == 1:
            print("\n------ Messages Limits --------")
            print("max queues system wide = %u" %values[0])
            print("max size of message (bytes) = %u" %values[1])
            print("default max size of queue (bytes) = %u" %values[2])
            print("\n------ Shared Memory Limits --------")
            print("max number of segments = %u" %values[3])
            print("max seg size (kbytes) = %u" %(values[4] / kb))
            print("max total shared memory (kbytes) = %u" %(values[5] * pagesize_kb))
            print("min seg size (bytes) = %u" %values[6])
            print("\n------ Semaphore Limits --------")
            print("max number of arrays = %u" %values[7])
            print("max semaphores per array = %u" %values[8])
            print("max semaphores system wide = %u" %values[9])
            print("max ops per semop call = %u" %values[10])
            print("semaphore max value = %u" %values[11])
        elif self.show_summary == 1:
            print("\n------ Messages Status --------")
            print("allocated queues = %u" %values[0])
            print("used headers = %u" %values[1])
            print("used space = %u  bytes" %values[2])
            print("\n------ Shared Memory Status --------")
            print("segments allocated   %u" %values[3])
            print("pages allocated    %u" %(values[4] / pagesize))
            print("pages resident   %u" %(values[5] / pagesize))
            print("pages swapped   %u" %values[6])
            print("Swap performance: %u attempts     %u successes" %(values[7], values[8]))
            print("\n------ Semaphore Status --------")
            print("used arrays = %u" %values[9])
            print("allocated semaphores = %u" %values[10])

    def connect(self):
        """ Establish a PMAPI context to archive, host or local, via args """
        self.context = pmapi.pmContext.fromOptions(self.opts, sys.argv)

if __name__ == '__main__':
    try:
        IPCS = Ipcs()
        IPCS.connect()
        IPCS.execute()
    except pmapi.pmErr as error:
        print("%s: %s" % (error.progname(), error.message()))
    except pmapi.pmUsageErr as usage:
        usage.message()
    except KeyboardInterrupt:
        pass
